src/solana/stake_account.ts

```diff
@@ -0,0 +1,162 @@
+// Copied from https://github.com/igneous-labs/solana-stake-sdk
+import { AccountInfo, PublicKey, ParsedAccountData } from '@solana/web3.js';
+import {
+  coerce,
+  create,
+  enums,
+  Infer,
+  instance,
+  nullable,
+  number,
+  string,
+  type,
+} from 'superstruct';
+import BigNumber from 'bignumber.js';
+
+export const PublicKeyFromString = coerce(
+  instance(PublicKey),
+  string(),
+  (value) => new PublicKey(value),
+);
+
+export class ParseStakeAccountError extends Error {}
+
+export const StakeState = {
+  Inactive: 'inactive',
+  Activating: 'activating',
+  Active: 'active',
+  Deactivating: 'deactivating',
+  Deactivated: 'deactivated',
+};
+
+export const StakeAccountType = enums([
+  'uninitialized',
+  'initialized',
+  'delegated',
+  'rewardsPool',
+]);
+
+export const BigNumFromString = coerce(
+  instance(BigNumber),
+  string(),
+  (value) => {
+    if (typeof value === 'string') return new BigNumber(value, 10);
+    throw new Error('invalid big num');
+  },
+);
+
+export type StakeMeta = Infer<typeof StakeMeta>;
+export const StakeMeta = type({
+  rentExemptReserve: BigNumFromString,
+  authorized: type({
+    staker: PublicKeyFromString,
+    withdrawer: PublicKeyFromString,
+  }),
+  lockup: type({
+    unixTimestamp: number(),
+    epoch: number(),
+    custodian: PublicKeyFromString,
+  }),
+});
+
+export type StakeAccountInfo = Infer<typeof StakeAccountInfo>;
+export const StakeAccountInfo = type({
+  meta: StakeMeta,
+  stake: nullable(
+    type({
+      delegation: type({
+        voter: PublicKeyFromString,
+        stake: BigNumFromString,
+        activationEpoch: BigNumFromString,
+        deactivationEpoch: BigNumFromString,
+        warmupCooldownRate: number(),
+      }),
+      creditsObserved: number(),
+    }),
+  ),
+});
+
+export type StakeAccount = Infer<typeof StakeAccount>;
+export const StakeAccount = type({
+  type: StakeAccountType,
+  info: StakeAccountInfo,
+});
+
+/**
+ * Converts an `AccountInfo<ParsedAccountData>` to an `AccountInfo<StakeAccount>`
+ * @param account raw accountinfo returned by getParsedProgramAccounts or getParsedAccountInfo
+ * @returns the parsed StakeAccount
+ * @throws ParseStakeAccountError if `account` is AccountInfo<Buffer> or if unable to parse account data
+ */
+export function parsedAccountInfoToStakeAccount({
+  executable,
+  owner,
+  lamports,
+  data,
+  rentEpoch,
+}: AccountInfo<Buffer | ParsedAccountData>): AccountInfo<StakeAccount> {
+  if (!('parsed' in data)) {
+    throw new ParseStakeAccountError(
+      'Raw AccountInfo<Buffer>, data not parsed',
+    );
+  }
+  try {
+    return {
+      executable,
+      owner,
+      lamports,
+      data: create(data.parsed, StakeAccount),
+      rentEpoch,
+    };
+  } catch (e) {
+    throw new ParseStakeAccountError((e as Error).message);
```

Comment:
It's better to avoid type casting is ts. 
Also, if we need to return some new errors and `throw this.handleError('UNSTAKE_ERROR', error)` doesn't fit here, so we need to consider a general solution for consistency.